Splitting of LAPACK routines into alloc/compute function pairs

jagot opened the pull request:
This makes it easy to allocate work/output arrays which can be reused for repeated calls to the compute function, as discussed in JuliaLang/LinearAlgebra.jl#328. In this PR, I have implemented this splitting for stev! and stegr!, with more to implemented along the way.

Commit:
Split stev! into alloc/compute routines as per #16263

Commit:
Split stegr! into alloc/compute routines as per #16263

andreasnoack commented:
Thanks for opening this. I don't think it should be necessary to write the `ccall` twice for each LAPACK routine and I think you should just merge the `_work` method into the method with few arguments. E.g. something like

```jl
function abc!(A::StridedMatrix, work::Vector, query::Bool, info::Ref{BlasInt})
    lwork = Ref(query ? -1 : length(work))
    ccall((:abs_, LAPACK), Void, (Ptr{T}, Ptr{T}, Ref{BlasInt}, Ref{BlasInt}),
        A, work, lwork, info)
    return A, info
end
function abc!(A::StridedMatrix)
    work = Array(T, 1)
    info = Ref{BlasInt}(0)
    abc!(A,work,true,info)
    @check info[]
    resize!(work, Int(work[1]))
    return abs!(A,work,false,info)
end
```

Would that work?

jagot commented:
@andreasnoack Sure, that would work, but I would personally prefer to not have to make the allocations myself. Maybe something like this?
```julia
function abc!(A::StridedMatrix, work::Vector, info::Ref{BlasInt}, query::Bool = false)
    lwork = Ref(query ? -1 : length(work))
    ccall((:abc_, LAPACK), Void, (Ptr{T}, Ptr{T}, Ref{BlasInt}, Ref{BlasInt}),
        A, work, lwork, info)
    return A, info
end
function abc!(A::StridedMatrix, query::Bool = false)
    work = Array(T, 1)
    info = Ref{BlasInt}(0)
    abc!(A,work,info,true)
    @check info[]
    resize!(work, Int(work[1]))
    query && return work,info
    return abc!(A,work,info)
end
```
One could then simply do
```julia
work,info = abc!(A, true)
abc!(A,work,info)
```

@tkelman What would be a sensible way to test this?

andreasnoack commented:
You'd need to return the workspace in the first function, right? Except for that, I agree. It would be better if  the function makes all the allocations and return them.

jagot committed:
Simplified stev! alloc/compute split

As per the discussion with @andreasnoack, the redundant extra _work
function is removed and is now part of the driver function.

jagot committed:
Simplified stegr! alloc/compute split

As per the discussion with @andreasnoack, the redundant extra _work
function is removed and is now part of the driver function.

jagot commented:
This pattern breaks, since type inference cannot say anything more than the result will be a tuple. Is it OK to change the tests to not perform static type inference and explicitly test the return types, depending on whether `query` is `true` or `false`?

KristofferC commented:
Won't your example be type unstable, that is, the return types will differ depending on values (not types) of the input arguments? If query is `true` you return a vector + info and if not you return a matrix + info. Isn't it better to factorize out the creation of working arrays to a separate function?

jagot commented:
@KristofferC Well, that is what I had in the first place, see 8573681, before the discussion with @andreasnoack.

martinholters commented:
If we want to stick with just one name (personally I'm doubtful about this), then the `Val{}` mechanism could be used for the `query` parameter to effectively have two type-stable methods - one for just allocating and one for allocating and computing (not returning the allocated data). (And then another one for computing without allocating, of course.)

I'd prefer having the allocation in a differently named function, though, or maybe put all this allocated data in a custom data type with a suitable constructor, so that the user can e.g. do
```Julia
res = abc!(A) # old style
```
or
```Julia
data = AbcData(size(A))
res = abc!(A, data) # no allocs here
```
where actually `abc!(A) = abc!(A, AbcData(size(A)))`.

Would that make sense?

jagot commented:
Yes, it could work, but it would require one type for every LAPACK function, and for those who return required work dimensions when in "query mode", the ccall would have to appear twice, once in the constructor and once in the compute method. Another point -- I don't know if this is a real concern, except for maybe really large problems -- is that if one uses a huge scratch area as work array for many different LAPACK calls, one would have to create these work objects manually anyway. I.e.
```julia
tmp = Array(Float64, 100000)

abc_data = AbcData(tmp)
def_data = DefData(tmp)

abc!(A, abc_data)
def!(D, def_data)
```

andreasnoack commented:
No `Val`s please. We should really try to avoid them.

I also don't think it would be worth the labor to define special types for the different subroutines in LAPACK. My experience is the subroutines are somewhat similar and then still a little different from eachother so it hard to abstract over all the subroutines. 

Wouldn't this version be fine?
```jl
function abc!(A::StridedMatrix, work::Vector, info::Ref{BlasInt}, query::Bool)
    lwork = Ref(query ? -1 : length(work))
    ccall((:abs_, LAPACK), Void, (Ptr{T}, Ptr{T}, Ref{BlasInt}, Ref{BlasInt}),
        A, work, lwork, info)
    return A, work, info
end
function abc!(A::StridedMatrix)
    work = Array(T, 1)
    info = Ref{BlasInt}(0)
    abc!(A,work,true,info)
    @check info[]
    resize!(work, Int(work[1]))
    return abs!(A,work,false,info)
end
```
?
Then the first run will give a work space of the right size and it is type stable. Here, you cannot run the one argument version only to get the allocations but is it a problem for your use case that e.g. you also compute the eigenvalues? I geuss it's the first thing you do with your newly allocated memory anyway.

Also, remember that you don't have to pass around `lwork` and `liwork` in Julia since it's just `length(work)` and `length(iwork)`.

jagot commented:
@andreasnoack _I_ would be fine with this, but it would break the interface, the driver always returning the work arrays as well. I am not willing to take upon myself to change all functions in Base dependent on every LAPACK function I edit :)

martinholters commented:
Right, changing the return types of the existing functions sounds like a really bad idea. Not only would it mean changing several places elsewhere in Julia - the worrying part is uses in packages which would somehow have to cope with return types depending on Julia version.

So let's summarize the required interfaces (and correct me if I'm wrong!):
* Old-style: All work area stuff done internally.
* Allocating: Allocates and returns the work areas; may or may not execute the algorithm, too.
* Non-allocating: Gets the work areas as arguments and executes the algorithm. Might for convenience return the work areas, too.

So the question is how to differentiate them. The signature of the first is pretty clear: It has to be as it is now. The signature of the last clearly differs because it has all the work area stuff as additional parameters. But first and second will likely require the same inputs, so I see the following options:
* Additional run-time parameter defaulting to evoke the old-style behavior (current PR status). This leads to type-instability so a no-go IMHO.
* Additional type parameter (`Val` or custom tag types). @andreasnoack doesn't like them; I'm skeptical, too.
* Different name (e.g. with `_prepare` or `_alloc` appended). Best one so far?

Once this is settled, one can ponder how this can be cleverly realized with as much code reuse as possible.

jagot commented:
I'd vote for an `_alloc` method; the intent is very clear from the name.

andreasnoack commented:
> Right, changing the return types of the existing functions sounds like a really bad idea.

I still think it's the best idea. Julia is right now version 0.4 and not >1.0 so we should focus on getting it right more than on avoiding any API breakage. We shouldn't break the API without a reason but the alternatives here are not good. Adding `_alloc` methods duplicates most of the code and the number of function names in `LAPACK`.

The LAPACK wrappers are not exported, shouldn't be and probably aren't used by that many end users. In most cases, if users find the higher-level linear algebra functions insufficient then we should improve them. Even those who believe they have to use the wrappers directly might not be seriously affected by such a change because the work space arrays would be among the last elements in the return tuple. Conditioning in on `VERSION` is also pretty easy. Finally, the extra changes to the Julia source are really minor. `stegr!` is used only five times and `stev!` only once and all six occurrences are in the same file. I believe that `stegr!` is one the LAPACK functions with the highest number of use cases in Base.

martinholters commented:
> Julia is right now version 0.4 and not >1.0 so we should focus on getting it right more than on avoiding any API breakage.

Yes, absolutely, but nevertheless quite often some effort is invested to have a deprecation/gradual adjustment strategy. And with the proposed return type change, I don't see it; adding a `@compat` for that looks quite hard, because the old function cannot be used to emulate the new one. If there is consensus that we shouldn't worry too much in this case, I'm perfectly fine with it.

>The LAPACK wrappers are not exported, shouldn't be and probably aren't used by that many end users. 

Err, but isn't the whole purpose of this PR to provide potentially more efficient LAPACK routines to end users/package developers? Or are there uses in Base that could profit from the alloc-once-run-many possibility?

>In most cases, if users find the higher-level linear algebra functions insufficient then we should improve them.

So do you see this PR as a preparation to rework the higher-level API to provide the alloc-once-run-many interface?

martinholters commented:
What about the following (assuming the original `abc!(A)` only returns the modified `A`):
```Julia
function abc!(A::StridedMatrix, work::Vector, info::Ref{BlasInt}, query::Bool=false)
    lwork = Ref(query ? -1 : length(work))
    ccall((:abc_, LAPACK), Void, (Ptr{T}, Ptr{T}, Ref{BlasInt}, Ref{BlasInt}),
        A, work, lwork, info)
    return A, work, info
end
function abc!(A::StridedMatrix, ::Void, ::Void)
    work = Array(T, 1)
    info = Ref{BlasInt}(0)
    abc!(A,work,info,true)
    @check info[]
    resize!(work, Int(work[1]))
    return abc!(A,work,info) # or maybe just "return A, work, info"?
end
function abc!(A::StridedMatrix)
    return abc!(A,nothing,nothing)[1]
end
```
One could even consider handling cases where some inputs are already allocated and the others are `nothing`.

andreasnoack commented:
Sorry, got distracted by other things. I have continued the discussion a bit offline and the consensus seems to be that we should go ahead and change the returned tuple if it simplifies the functions which I think it does in this case. @jagot I still think the general idea in this PR is great so I hope you are okay with continuing in this direction..

Regarding @martinholters comments
> I don't see it; adding a @compat for that looks quite hard

I think that `@compat` is mainly important for very common functions. Users who are able to use LAPACK functions would probably also easily be able to add a `VERSION` condition to their code id necessary (it's still not completely clear it will be).

> Err, but isn't the whole purpose of this PR to provide potentially more efficient LAPACK routines to end users/package developers? Or are there uses in Base that could profit from the alloc-once-run-many possibility?

It is, but it should only be a minority of our users who'd need to call the wrappers directly. For most users, the cost of allocation temporary storage is small compared to the linear algebra operation.

> So do you see this PR as a preparation to rework the higher-level API to provide the alloc-once-run-many interface?

Not really. As mentioned above, I think that these costs are insignificant for most users.

JaredCrean2 commented:
> I think that these costs are insignificant for most users.

To get C like performance, getting rid of the memory allocation is essential.  For a high-performance language to not have an efficient Lapack interface is really a problem.

martinholters commented:
> To get C like performance, getting rid of the memory allocation is essential. For a high-performance language to not have an efficient Lapack interface is really a problem.

Yes, and we all agree we want the allocation-free interface, we're only debating what exactly it should look like. And for the non-power-users, we keep an easy to use higher-level interface, which alas is not maximally performant due to its memory allocations.

If the consensus is that changing the return type is OK because there currently are only very few direct uses of the LAPACK routines out there (and after a brief search I agree that seems to be the case indeed), then go ahead.

jagot commented:
Ok, I will go ahead. Just for curiosity, is returning a value a no-op, if the value is not caught by the caller?

andreasnoack commented:
@jagot Any progress here?

jagot commented:
Sorry, I have been preoccupied with research for quite some time. I tried to implement the splitting the way we discussed above, but could not achieve type stability. I will see if I can get around it again this week.

ViralBShah commented:
Is there still interest in seeing this through? It is a nice idea - but should be done across the board. Also may be easier to do if we were able to put some of the linear algebra into its own external packges.

chriscoey commented:
This would be great!

jagot commented:
Last time I looked at this (a couple of years ago), I could not really achieve type stability (I was very green at the time). Later, I implemented it for the routine I actually needed ([`stegr!`](https://github.com/JuliaDiffEq/ExponentialUtilities.jl/blob/master/src/StegrWork.jl)). Recently, @andreasnoack seemed to be of the opinion that it would be a lot of work, for marginal gain. In my case, where I want/need allocation-free time-stepping routines, it's surely nice, but I don't really have the time to engineer the rewrite of the whole BLAS library.

Could one be clever with some metaprogramming, to automatically specify the different interfaces (allocating/non-allocating with work `struct`)?

ViralBShah commented:
I do agree that the effort would be substantial. It is certainly easy to call any necessary lapack routines directly for ones own work.

ViralBShah commented:
I'm going to close this one for now.
